### PR TITLE
Make Layout's align a NonZeroUsize

### PR DESCRIPTION
**Pull request by gnzlbg**
This PR makes the `Layout`'s align field a `NonZeroUsize` since it cannot ever be zero, not even while building a `Layout`. It also contains some drive-by minor cleanups over the docs and the code, like updating the documented error types, or using the `size()` and `align()` methods instead of accessing the fields directly (the latter was required for the `NonZeroUsize` change anyways).

r? @SimonSapin 

cc @Amanieu

### CODE REVIEW

**Review by SimonSapin — APPROVED** [review]
Looks good, thanks! Just one thing:

File: src/libcore/alloc.rs [hunk 1]
```diff
@@ -203,9 +201,12 @@ impl Layout {
     /// Note that the utility of the returned value requires `align`
     /// to be less than or equal to the alignment of the starting
     /// address for the whole allocated block of memory. One way to
-    /// satisfy this constraint is to ensure `align <= self.align`.
+    /// satisfy this constraint is to ensure `align <= self.align()`.
     #[inline]
     pub fn padding_needed_for(&self, align: usize) -> usize {
+        // **FIXME**: This function is only called with proper power-of-two
+        // alignments. Maybe we should turn this into a real assert!.
+        debug_assert!(align.is_power_of_two());
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply src/libcore/alloc.rs]: It is *not* undefined behavior to call this method with a non-power-of-two, that only causes it to return an integer value that is not meaningful. So it shouldn’t be `unsafe` IMO. Calling it a precondition really depends on what you mean by that.

[the remark this answers is not in the document]
Comment (gnzlbg) [reply src/libcore/alloc.rs]: > It is not undefined behavior to call this method with a non-power-of-two, that only causes it to return an integer value that is not meaningful.

It is undefined if we say it is. That allows us to return a meaningless value, or assert, or abort, or... Maybe I meant implementation defined here, but for implementation defined we need to define what the behavior actually is. With undefined we can do anything, including returning a meaningless value.

> Calling it a precondition really depends on what you mean by that.

A precondition on the result having any meaning. AFAICT most APIs in std use `Option::None` to indicate that a result is meaningless, but maybe there are some APIs already that just return meaningless `usize`s already?

In any case, as long `std` tests are built in debug mode and run by some build bot, the `debug_assert` should tell us if this behavior is relied upon anywhere in std or not.

[the remark this answers is not in the document]
Comment (SimonSapin) [reply src/libcore/alloc.rs]: > It is undefined if we say it is.

Sure, but why would we? This method only does integer arithmetic.

Anyway, I’m ready to r+ this PR as soon as this chunk is removed. Maybe the design or existence of this method should be revisited, but in the middle of an unrelated PR is not the place to do it.

**Comment by gnzlbg** [comment]
@SimonSapin I’ll remove the debug assert and ask about it in the tracking
issue.

**Pull request merged**